pkg/api/controller_test.go

```diff
@@ -1549,3 +1550,150 @@ func TestController_MergeDiffWithParent(t *testing.T) {
 		t.Fatal("Diff results not as expected:", diff)
 	}
 }
+
+func TestController_MergeIntoStage(t *testing.T) {
+	clt, deps := setupClientWithAdmin(t, "")
+	ctx := context.Background()
+	// setup env
+	repo := testUniqueRepoName()
+	_, err := deps.catalog.CreateRepository(ctx, repo, onBlock(deps, repo), "main")
+	testutil.Must(t, err)
+	_, err = deps.catalog.CreateBranch(ctx, repo, "branch1", "main")
+	testutil.Must(t, err)
+	testutil.MustDo(t, "create entry bar1", deps.catalog.CreateEntry(ctx, repo, "branch1", catalog.DBEntry{Path: "foo/bar1", PhysicalAddress: "bar1addr", CreationDate: time.Now(), Size: 1, Checksum: "cksum1"}))
+	commit1, err := deps.catalog.Commit(ctx, repo, "branch1", "some message", DefaultUserID, nil)
+	testutil.Must(t, err)
+
+	// merge 'branch1' to 'main' staging
+	commitMessage := "add foo1"
+	tagResp, err := clt.MergeIntoBranchWithResponse(ctx, repo, "branch1", "main$", api.MergeIntoBranchJSONRequestBody{
+		Message: &commitMessage,
+	})
+	testutil.Must(t, err)
+	if tagResp.JSONDefault == nil {
+		t.Fatalf("merge to staging should fail with an error, got: %v", tagResp)
+	}
+
+	// merge 'branch1' to 'main' committed
+	tagResp2, err := clt.MergeIntoBranchWithResponse(ctx, repo, "branch1", "main", api.MergeIntoBranchJSONRequestBody{
+		Message: &commitMessage,
+	})
+	verifyResponseOK(t, tagResp2, err)
+
+	commit, err := clt.GetCommitWithResponse(ctx, repo, tagResp2.JSON200.Reference)
+	verifyResponseOK(t, commit, err)
+	if len(commit.JSON200.Parents) != 2 || commit.JSON200.Parents[1] != commit1.Reference {
+		t.Fatal("Merge parents should include source commit", commit.JSON200)
+	}
+}
+
+func TestController_CreateTag(t *testing.T) {
+	clt, deps := setupClientWithAdmin(t, "")
+	ctx := context.Background()
+	// setup env
+	repo := testUniqueRepoName()
+	_, err := deps.catalog.CreateRepository(ctx, repo, onBlock(deps, repo), "main")
+	testutil.Must(t, err)
+	testutil.MustDo(t, "create entry bar1", deps.catalog.CreateEntry(ctx, repo, "main", catalog.DBEntry{Path: "foo/bar1", PhysicalAddress: "bar1addr", CreationDate: time.Now(), Size: 1, Checksum: "cksum1"}))
+	commit1, err := deps.catalog.Commit(ctx, repo, "main", "some message", DefaultUserID, nil)
+	testutil.Must(t, err)
+
+	t.Run("ref", func(t *testing.T) {
+		tagResp, err := clt.CreateTagWithResponse(ctx, repo, api.CreateTagJSONRequestBody{
+			Id:  "tag1",
+			Ref: commit1.Reference,
+		})
+		verifyResponseOK(t, tagResp, err)
+	})
+
+	t.Run("branch", func(t *testing.T) {
+		tagResp, err := clt.CreateTagWithResponse(ctx, repo, api.CreateTagJSONRequestBody{
+			Id:  "tag2",
+			Ref: "main",
+		})
+		verifyResponseOK(t, tagResp, err)
+	})
+
+	t.Run("explicit_branch_latest", func(t *testing.T) {
```

Comment:
a suggestion - you can use "branch_with_latest_modifier" or "branch_with_committed_modifier". IMHO, explicit is not clear enough here